packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Spec](./README.md)
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text data-key="name"></nimble-table-column-text>
+    <nimble-table-column-icon data-key="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation. The ordering of the columns in the markup will determine the visual ordering of the columns (top to bottom equals left to right...unless in 'rtl'). Re-ordering of columns will be done, at least at first, through the re-ordering of the columns elements in the DOM.
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    /*
+     * @internal
+     */
+    @observable
+    public readonly slottedColumns: TableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where the `TCellData` type is a superset of the type represented by [`TableRecord`](https://github.com/ni/nimble/blob/3e4b8d3dd59431d1671e381aa66052db57bc475c/packages/nimble-components/src/table/types.ts#L24):
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `TableColumn<>`
+
+This abstract class is what a column web component (i.e. a slotted column) must extend.
+
+```TS
+abstract class TableColumn<TCellData extends TableRowData = unknown> {
```

Comment:
I acknowledge that this is just a spec, but you won't be able to assign `unknown` as the default value here because `unknown` isn't assignable to `TableRowData`. This also needs to be updated to use `TableRecord`.

```suggestion
abstract class TableColumn<TCellData extends TableRecord = TableRecord> {
```